### PR TITLE
refactor(Raster): use extent to handle texture instead of index integer.

### PR DESCRIPTION
**Pull request by gchoqueux**
## Description
Use extent to handle texture instead of index integer.

## Motivation and Context
it's more appropriate and intuitive to apply a texture on extent.
In addition, one texture could be applied individually.

### CODE REVIEW

**Comment by zarov** [comment]
I got some bad display when getting closer to the ground:

Screenshot from 2019-10-11 11-02-58

While I should get:

Screenshot from 2019-10-11 11-03-15

File: src/Renderer/MaterialLayer.js [hunk 1]
```diff
@@ -82,8 +82,9 @@ class MaterialLayer {
         defineLayerProperty(this, 'zmin', layer.zmin, defaultEle.zmin);
         defineLayerProperty(this, 'zmax', layer.zmax, defaultEle.zmax);
 
-        this.textures = [];
-        this.offsetScales = [];
+        this.textures = new Map();
+        extents.forEach(ex => this.textures.set(ex, null));
+        this.offsetScales = new Map();
```
Comment (zarov) [inline src/Renderer/MaterialLayer.js]: You don't set `this.offsetScales` ?

File: src/Renderer/MaterialLayer.js [hunk 2]
```diff
@@ -126,22 +130,30 @@ class MaterialLayer {
             }
         }
         this.level = EMPTY_TEXTURE_ZOOM;
-        this.textures = [];
-        this.offsetScales = [];
+        this.textures.forEach((v, k) => this.textures.set(k, null));
```
Comment (zarov) [inline src/Renderer/MaterialLayer.js:132]: No `this.textures.clear()` instead ?

File: src/Renderer/MaterialLayer.js [hunk 3]
```diff
@@ -126,22 +130,30 @@ class MaterialLayer {
...
         this.material.layersNeedUpdate = true;
     }
 
-    setTexture(index, texture, offsetScale) {
-        this.level = (texture && (index == 0)) ? texture.coords.zoom : this.level;
-        this.textures[index] = texture || null;
-        this.offsetScales[index] = offsetScale;
+    setTexture(texture, extent) {
+        this.level = texture ? texture.coords.zoom : this.level;
+        this.textures.set(extent, texture || null);
```
Comment (zarov) [inline src/Renderer/MaterialLayer.js:138]: I'm not a fan of using `extent` as a key, maybe it's me but using such a "complex" key when getting something in a `Map` seems to be slow and overkill. Why not write a method that generate a unique key from the content of the extent ?

However, I see two drawbacks in this suggestion (that needs to be verified):
- the conversion to a string can mean a lot of process if it is called a lot
- I'm not sure if the key is used as it is sometimes (it seems in `updateLayeredMaterialNodeImagery`, so...)

**Review by zarov — COMMENTED** [review]
Looking good, can you do a quick benchmark to see the impact on performance ? I'm quite sceptical about this, sorry. I'm afraid it will have too much of an impact.

File: src/Renderer/MaterialLayer.js (hunk 1 above)
Comment (gchoqueux) [reply src/Renderer/MaterialLayer.js]: No, the previous `offsetScales` wasn't initializing.

File: src/Renderer/MaterialLayer.js (hunk 2 above)
Comment (gchoqueux) [reply src/Renderer/MaterialLayer.js:132]: I would like keep this `Extents` to assign the texture by `Extent`

File: src/Renderer/MaterialLayer.js (hunk 3 above)
Comment (gchoqueux) [reply src/Renderer/MaterialLayer.js:138]: Me too, I'm not fan. We can optimize later.

File: src/Renderer/MaterialLayer.js (hunk 1 above)
Comment (gchoqueux) [reply src/Renderer/MaterialLayer.js]: I thinks remove this property.

File: src/Renderer/MaterialLayer.js (hunk 3 above)
Comment (zarov) [reply src/Renderer/MaterialLayer.js:138]: Good to me

File: src/Renderer/MaterialLayer.js (hunk 1 above)
Comment (gchoqueux) [reply src/Renderer/MaterialLayer.js]: done

File: src/Layer/TiledGeometryLayer.js [hunk 4]
```diff
@@ -371,14 +372,16 @@ class TiledGeometryLayer extends GeometryLayer {
...
-            const currentTexture = nodeLayer.textures[0];
-            if (currentTexture && currentTexture.extent) {
-                const offsetScale = nodeLayer.offsetScales[0];
-                const ratio = offsetScale.z;
-                // ratio is node size / texture size
-                if (ratio < 1 / Math.pow(2, this.maxDeltaElevationLevel)) {
-                    return false;
+            for (const [extent, texture] of nodeLayer.textures.entries()) {
+                if (texture) {
+                    extent.offsetToParent(texture.coords, offsetScale);
```
Comment (zarov) [inline src/Layer/TiledGeometryLayer.js:377]: You can avoid calling `offsetToParent` by maybe adding a new method on `Extent`, something that only compute the `z` part, as it is the only thing used here.

**Comment by Desplandis** [comment]
@gchoqueux Are you still working on it?

**Pull request closed without merging**